content/300-guides/500-other/900-advanced-database-tasks/06-typesafe-raw-sql/01-safeql.mdx

```diff
@@ -0,0 +1,369 @@
+
+```terminal
+npm install dotenv
+```
+
+Then update your ESLint config as follows:
+
+```js file=.eslintrc.js highlight=1,6-9,16;add
+require('dotenv').config()
+
+/** @type {import('eslint').Linter.Config} */
+module.exports = {
+  plugins: ['@ts-safeql/eslint-plugin'],
+  // exclude `parserOptions` if you are not using TypeScript
+  parserOptions: {
+    project: './tsconfig.json',
+  },
+  rules: {
+    '@ts-safeql/check-sql': [
+      'error',
+      {
+        connections: [
+          {
+            connectionUrl: process.env.DATABASE_URL,
+            // The migrations path:
+            migrationsDir: './prisma/migrations',
+            targets: [
+              // what you would like SafeQL to lint. This makes `prisma.$queryRaw` and `prisma.$executeRaw`
+              // commands linted
+              { tag: 'prisma.+($queryRaw|$executeRaw)', transform: '{type}[]' },
+            ],
+          },
+        ],
+      },
+    ],
+  },
+}
+```
+
+SafeQL is now fully configured to help you write better raw SQL using Prisma Client.
+
+## 4. Creating extensions to make raw SQL queries type-safe
+
+In this section, we'll create two `model` extensions with custom queries to be able to work conveniently with the `PointOfInterest` model:
+
+1. A `create` query that allows us to create new `PointOfInterest` records in the database
+1. A `findClosestPoints` query that returns the `PointOfInterest` records that are closest to a given coordinate
+
+### 4.1. Adding an extension to create `PointOfInterest` records
+
+The `PointOfInterest` model in the Prisma schema uses an `Unsupported` type. As a consequence, the generated `PointOfInterest` type in Prisma Client can't be used to carry values for latitude and longitude.
+
+We will resolve this by defining two custom types that better represent our model in TypeScript:
+
+```ts
+type MyPoint = {
+  latitude: number
+  longitude: number
+}
+
+type MyPointOfInterest = {
+  name: string
+  location: Point
+}
+```
+
+Next, you can add a `create` query to the `pointOfInterest` property of your Prisma Client:
+
+```ts highlight=19;normal
+const prisma = new PrismaClient().$extends({
+  model: {
+    pointOfInterest: {
+      async create(data: {
+        name: string
+        latitude: number
+        longitude: number
+      }) {
+        // Create an object using the custom types from above
+        const poi: MyPointOfInterest = {
+          name: data.name,
+          location: {
+            latitude: data.latitude,
+            longitude: data.longitude,
+          },
+        }
+
+        // Insert the object into the database
+        const point = `POINT(${poi.location.longitude} ${poi.location.latitude})`
+        await prisma.$queryRaw`
+          INSERT INTO "PointOfInterest" (name, location) VALUES (${poi.name}, ST_GeomFromText(${point}, 4326));
+        `
+
+        // Return the object
+        return poi
+      },
+    },
+  },
+})
```

Comment:
Is this a valid query? It seems we're returning an object we created rather than the response from the query?  What happens if the query returns an error? I think we should return the value from the query at least to make this a valid query

Reply:
> Is this a valid query?

Yes it is 👍 

> It seems we're returning an object we created rather than the response from the query?

Correct! `INSERT` doesn't return any data so we're indeed returning the plain JS object here. Tbh, I was a bit confused by this at first but now think this is the most pragmatic way of doing things in this article.

> What happens if the query returns an error?

I'd argue that error handling is beyond the scope of this article.

> I think we should return the value from the query at least to make this a valid query

It is a valid query and the query doesn't return anything.